pkg/provisioning/service.go

```diff
@@ -234,54 +237,30 @@ func (s *Service) collectConnectorStates(ctx context.Context, pl *pipeline.Insta
 			return nil, fmt.Errorf("could not get connector %q: %w", connID, err)
 		}
 
-		switch oldConn := oldConn.(type) {
-		case connector.Destination:
-			oldState := oldConn.State()
-			connStates[connID] = func(ctx context.Context) error {
-				newConn, err := s.connectorService.Get(ctx, connID)
-				if err != nil {
-					return fmt.Errorf("could not get connector %q: %w", connID, err)
-				}
-				if newConn.Config().Plugin != oldConn.Config().Plugin {
-					s.logger.Warn(ctx).
-						Str(log.PluginNameField, newConn.Config().Plugin).
-						Msg("plugin name changed, could not apply old destination state, the connector will start from the beginning")
-					return nil
-				}
-				_, err = s.connectorService.SetDestinationState(ctx, connID, oldState)
-				if cerrors.Is(err, connector.ErrInvalidConnectorType) {
-					s.logger.Warn(ctx).
-						Str(log.ConnectorIDField, connID).
-						Err(err).
-						Msg("could not apply old destination state, the connector will start from the beginning")
-					return nil
-				}
-				return err
+		oldState := oldConn.State
+		if oldState == nil {
+			continue // no need to copy state
+		}
+		connStates[connID] = func(ctx context.Context) error {
+			newConn, err := s.connectorService.Get(ctx, connID)
+			if err != nil {
+				return fmt.Errorf("could not get connector %q: %w", connID, err)
 			}
-		case connector.Source:
-			oldState := oldConn.State()
-			connStates[connID] = func(ctx context.Context) error {
-				newConn, err := s.connectorService.Get(ctx, connID)
-				if err != nil {
-					return fmt.Errorf("could not get connector %q: %w", connID, err)
-				}
-				if newConn.Config().Plugin != oldConn.Config().Plugin {
-					s.logger.Warn(ctx).
-						Str(log.PluginNameField, newConn.Config().Plugin).
-						Str(log.ConnectorIDField, connID).
-						Msg("plugin name changed, could not apply old source state, the connector will start from the beginning")
-					return nil
-				}
-				_, err = s.connectorService.SetSourceState(ctx, connID, oldState)
-				if cerrors.Is(err, connector.ErrInvalidConnectorType) {
-					s.logger.Warn(ctx).
-						Str(log.ConnectorIDField, connID).
-						Err(err).
-						Msg("could not apply old source state, the connector will start from the beginning")
-					return nil
-				}
-				return err
+			if newConn.Plugin != oldConn.Plugin {
+				s.logger.Warn(ctx).
+					Str(log.PluginNameField, newConn.Plugin).
+					Msg("plugin name changed, could not apply old destination state, the connector will start from the beginning")
+				return nil
 			}
+			_, err = s.connectorService.SetState(ctx, connID, oldState)
+			if cerrors.Is(err, connector.ErrInvalidConnectorType) {
+				s.logger.Warn(ctx).
+					Str(log.ConnectorIDField, connID).
+					Err(err).
+					Msg("could not apply old state, the connector will start from the beginning")
+				return nil
```

Comment:
WDYT about failing in all cases where the state cannot be set? Otherwise, a connector would start from the beginning (and maybe it's a huge amount of data) due to an unintentional mistake (e.g. a connector was changed from source to destination by mistake).

Reply:
I don't think we should do that, essentially this would mean we would reject a valid pipeline configuration. We said that whatever is in the pipeline config file should be created in Conduit, regardless of what was there before, that's why we only warn the user that they might have done something unexpected, but we should not stop the pipeline from being updated.

Reply:
You have a point there :100: . My biggest worry is that this error can cause quite some trouble (re-reading all the data from a source that is maybe a high-volume one), and the change could have been an unintentional one. IMHO the consequences of re-reading a source are big enough to be strict about handling this.

In other words, even though a pipeline config is valid (and we would accept is as such), the fact that the state cannot be set means an "bad" change happened.

Reply:
> because the same pipeline config file could be accepted by one Conduit instance and not by another.

Right, that adds a lot of weight to the argument you mentioned. Given that this was already discussed, I wouldn't like to start a rediscussion, especially not here, so I'll only ask if we have this documented somewhere, and if not, to ask to add a comment somewhere explaining the reasoning behind this decision?

Reply:
I gathered the decisions in an ADR: https://github.com/ConduitIO/conduit/pull/827